### PR TITLE
First version of solution

### PR DESCRIPTION
**Pull request by MorozVladislav**
Seems to be working without unhandled exceptions. It was quite an interesting thing to dig in the Python internal structures and change their behaviour.

### CODE REVIEW

File: solutions/task_2.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+"""
+TASK: Create 'Fake' class to run code below without errors.
+"""
+
+import random
+import string
+
+
+class Fake(dict):
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: `dict`? Why do you need it?

File: solutions/task_2.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    def __getattr__(self, name):
+        print("Non existing attribute {} has been called".format(name))
+        return self
+
+    def __call__(self, *args, **kwargs):
+        try:
+            dict.__call__(*args, **kwargs)
+            return self
+        except (TypeError, ValueError):
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: As I understand all the code below is just for you (for your experiments with args and kwargs). Would you like to leave this code in the solution?

Comment (dmitry-moroz) [reply solutions/task_2.py]: Anyway I decided to review this code too...

File: solutions/task_2.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            print("The attribute has been called with:")
+            if len(args) > 0:
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: `if args:`

File: solutions/task_2.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+                print("- positional arguments: {}".format(str(", ".join(args))))
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: 1) `join` returns string, why do you use `str()` here?
2) I think it's enough: 
`print("- positional arguments: {}".format(args))`

Comment (dmitry-moroz) [reply solutions/task_2.py]: All standard objects have got `__str__` method, so it allows you pass them to `format `method when you are formatting string

File: solutions/task_2.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            else:
+                print("- no positional arguments")
+            if len(kwargs) > 0:
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: `if kwargs:`

File: solutions/task_2.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                kwargs_list = []
+                for k, v in kwargs.items():
+                    kwargs_list.append(str(k) + " = " + str(v))
+                print("- keyword arguments: {}".format(str(", ".join(kwargs_list))))
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: I think it's enough:
`print("- keyword arguments: {}".format(kwargs))`

File: solutions/task_2.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            else:
+                print("- no keyword arguments")
+            return self
+
+    def __getitem__(self, item):
+        try:
+            return dict.__getitem__(self, item)
+        except KeyError:
+            print("Unexisting key {} has been called".format(item))
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: > has been **called**

Really?

File: solutions/task_2.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                for k, v in kwargs.items():
+                    kwargs_list.append(str(k) + " = " + str(v))
+                print("- keyword arguments: {}".format(str(", ".join(kwargs_list))))
+            else:
+                print("- no keyword arguments")
+            return self
+
+    def __getitem__(self, item):
+        try:
+            return dict.__getitem__(self, item)
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: **Important thing:**
In this task you have redefined behavior of some methods from parent class. -> So, for the first you a tying (`try:`) to behave in normal way (like parent object should behave). But you don't know how parent method should behave. Here you are truing to guess it 👎 

So if you are not Vanga you have to call this method from parent object. To do it right read about `super`.

File: solutions/task_2.py (hunk 5 above)
Comment (dmitry-moroz) [reply solutions/task_2.py]: If you don't understand why - read and then ask.

File: solutions/task_2.py (hunk 8 above)
Comment (dmitry-moroz) [reply solutions/task_2.py]: Python3: https://docs.python.org/3.5/library/functions.html#super
Python2: https://docs.python.org/2/library/functions.html#super
Python 2 notation is also supported with python 3.

**Comment by MorozVladislav** [comment]
I'm going to bed now. If some mistakes are found please comment on them.

**Comment by dmitry-moroz** [comment]
I have to repeat: try to find 5 minutes to read about super and use it right..

File: solutions/task_2.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+TASK: Create 'Fake' class to run code below without errors.
+"""
+
+import random
+import string
+
+
+class Fake(object):
+    def __getattr__(self, name):
+        print("Non existing attribute {} has been called".format(name))
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: `called`
Is it to hard?

File: solutions/task_2.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        return self
+
+    def __call__(self, *args, **kwargs):
+        try:
+            super().__call__(self, *args, **kwargs)
+            return self
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: Why are you returning `self `here?

File: solutions/task_2.py (hunk 9 above)
Comment (dmitry-moroz) [reply solutions/task_2.py]: Where is `name()` **call**?

File: solutions/task_2.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+                print("- positional arguments: {}".format(args))
+            else:
+                print("- no positional arguments")
+            if kwargs:
+                print("- keyword arguments: {}".format(kwargs))
+            else:
+                print("- no keyword arguments")
+            return self
+
+    def __getitem__(self, item):
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: Are you going to add ability to get existing item? (to be similar with `__call__`)

**Comment by dmitry-moroz** [comment]
Suggestion: stop solving non-exciting task, get rid of all `try `constructions and unnecessary code, let your `Fake `class to be fake and that is all. (if you wish of cause)

File: solutions/task_2.py [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class Fake(object):
+    def __getattr__(self, name):
+        print("Non existing attribute {} has been called".format(name))
+        return self
+
+    def __call__(self, *args, **kwargs):
+        try:
+            super().__call__(self, *args, **kwargs)
```
Comment (dmitry-moroz) [inline solutions/task_2.py]: I am forced to repeat: try to find 5 min to read about super and use it right.

**Review by dmitry-moroz — APPROVED** [review]

**Pull request merged**